tests/nosetests/pyanaconda_tests/dbus_client_test.py

```diff
@@ -0,0 +1,441 @@
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
+# Public License for more details.  You should have received a copy of the
+# GNU General Public License along with this program; if not, write to the
+# Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+# 02110-1301, USA.  Any Red Hat trademarks that are incorporated in the
+# source code or documentation are not subject to the GNU General Public
+# License and may only be used or replicated with the express permission of
+# Red Hat, Inc.
+#
+# Red Hat Author(s): Vendula Poncova <vponcova@redhat.com>
+#
+import unittest
+from textwrap import dedent
+from unittest.mock import Mock, patch
+
+from pyanaconda.core.signal import Signal
+from pyanaconda.dbus.client import ClientObjectHandler, ObjectProxy, GLibClient
+from pyanaconda.dbus.constants import DBUS_FLAG_NONE, DBUS_DEFAULT_TIMEOUT
+from pyanaconda.dbus.error import ErrorRegister
+from pyanaconda.dbus.specification import DBusSpecification
+from pyanaconda.dbus.typing import get_variant, get_variant_type, VariantType
+
+import gi
+gi.require_version("Gio", "2.0")
+from gi.repository import Gio
+
+
+class FakeException(Exception):
+    """Fake exception from DBus calls."""
+    pass
+
+
+class VariantTypeFactory(object):
+    """Return same objects for same type strings.
+
+    This factory allows to easily test calls with variant types.
+    """
+
+    def __init__(self):
+        self._types = {}
+        self._new = VariantType.new
+        self._str = VariantType.__str__
+        self._repr = VariantType.__repr__
+
+    def set_up(self):
+        VariantType.new = self.get
+        VariantType.__str__ = VariantType.dup_string
+        VariantType.__repr__ = VariantType.dup_string
+
+    def tear_down(self):
+        VariantType.new = self._new
+        VariantType.__str__ = self._str
+        VariantType.__repr__ = self._repr
+
+    def get(self, type_string):
+        return self._types.setdefault(type_string, self._new(type_string))
+
+
+class DBusClientTestCase(unittest.TestCase):
+    """Test DBus clinet support."""
+
+    NO_REPLY = get_variant("()", ())
+
+    def setUp(self):
+        self.maxDiff = None
+        self.message_bus = Mock()
+        self.connection = self.message_bus.connection
+        self.service_name = "my.service"
+        self.object_path = "/my/object"
+        self.handler = None
+        self.proxy = None
+
+        self.variant_type_factory = VariantTypeFactory()
+        self.variant_type_factory.set_up()
+
+    def tearDown(self):
+        self.variant_type_factory.tear_down()
+
+    def variant_type_factory_test(self):
+        """Test the variant type factory."""
+        self.assertEqual(str(get_variant_type("s")), "s")
+        self.assertEqual(repr(get_variant_type("i")), "i")
+
+        self.assertEqual(get_variant_type("s"), get_variant_type("s"))
+        self.assertEqual(get_variant_type("i"), get_variant_type("i"))
```

Comment:
Is this test of idempotence? If yes, please write a comment.

Reply:
I don't know. It just tests that `get_variant_type` returns same objects for the same attributes to confirm that the mocked `VariantType` works as expected. Otherwise, you can get some very ugly failures from the other tests.

Reply:
OK, you can leave it here. I'm telling myself that working in the functional languages has its benefits :D.